File: app/src/main/java/com/futsch1/medtimer/medicine/ReminderViewAdapter.java
package com.futsch1.medtimer.medicine;


import android.os.HandlerThread;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.DiffUtil;

import com.futsch1.medtimer.database.Medicine;
import com.futsch1.medtimer.database.Reminder;
import com.futsch1.medtimer.helpers.IdlingListAdapter;

public class ReminderViewAdapter extends IdlingListAdapter<Reminder, ReminderViewHolder> {

    private final FragmentActivity fragmentActivity;
    private final HandlerThread thread;
    private Medicine medicine;

    public ReminderViewAdapter(FragmentActivity fragmentActivity, HandlerThread thread) {
        super(new ReminderDiff());
        this.fragmentActivity = fragmentActivity;
        this.thread = thread;
        setHasStableIds(true);
    }

    public void setMedicine(Medicine medicine) {
        this.medicine = medicine;
    }


    // Create new views (invoked by the layout manager)
    @NonNull
    @Override
    public ReminderViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return ReminderViewHolder.create(parent, fragmentActivity, thread);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NonNull ReminderViewHolder holder, final int position) {
        Reminder current = getItem(position);
        holder.bind(current, medicine);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).reminderId;
    }

    public static class ReminderDiff extends DiffUtil.ItemCallback<Reminder> {

        @Override
        public boolean areItemsTheSame(@NonNull Reminder oldItem, @NonNull Reminder newItem) {
            return oldItem == newItem;
        }

        @Override
        public boolean areContentsTheSame(@NonNull Reminder oldItem, @NonNull Reminder newItem) {
            return oldItem.reminderId == newItem.reminderId && oldItem.amount.equals(newItem.amount)
                    && oldItem.timeInMinutes == newItem.timeInMinutes;
        }
    }
}

File: app/src/main/java/com/futsch1/medtimer/medicine/ReminderViewAdapter.kt
package com.futsch1.medtimer.medicine

import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.DiffUtil
import com.futsch1.medtimer.helpers.IdlingListAdapter
import com.futsch1.medtimer.model.Medicine
import com.futsch1.medtimer.model.Reminder
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject


class ReminderViewAdapter @AssistedInject constructor(
    @Assisted private val fragmentActivity: FragmentActivity,
    private val reminderViewHolderFactory: ReminderViewHolder.Factory
) : IdlingListAdapter<Reminder, ReminderViewHolder>(ReminderDiff()) {
    @AssistedFactory
    interface Factory {
        fun create(fragmentActivity: FragmentActivity): ReminderViewAdapter
    }

    private lateinit var medicine: Medicine

    init {
        setHasStableIds(true)
    }

    fun setMedicine(medicine: Medicine) {
        this.medicine = medicine
    }


    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReminderViewHolder {
        return reminderViewHolderFactory.create(parent, fragmentActivity)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ReminderViewHolder, position: Int) {
        val current = getItem(position)
        holder.bind(current, medicine)
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).id.toLong()
    }

    class ReminderDiff : DiffUtil.ItemCallback<Reminder>() {
        override fun areItemsTheSame(oldItem: Reminder, newItem: Reminder): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Reminder, newItem: Reminder): Boolean {
            return oldItem == newItem
        }
    }
}

